### ISSUE TITLE
Installer places downloaded multisite projects in root folders

### ISSUE THREAD

**Issue by docwilmot** [backdrop/backdrop-issues#3274]
## Describe your issue or idea
In multisite, if you install a module with the project installer, it goes into the root`/modules` folder, instead of the `sites/sitefolder/modules` folder as expected.

**Comment by BWPanda** [comment]
Yes, I just started playing with Project Installer for the first time and saw this.

**Comment by docwilmot** [comment]
This is intentional apparently, same in Drupal: https://api.backdropcms.org/api/backdrop/core%21modules%21system%21system.updater.inc/function/ModuleUpdater%3A%3AgetInstallDirectory/1

@BWPanda think we should change this? Or close this issue? I guess it makes sense that contrib project should go into their main folder; only really custom projects *need* to go into the multisite folders. Unless anyone objects...

**Comment by BWPanda** [comment]
I object. Putting projects in the root directories is for when they're used for _all_ sites. But why assume that's the case here? Why assume that installing a module from the UI of a specific website means that module is going to be used on other websites?

I think if we're not willing to change this outright, we should at least provide an option to specify where new modules are installed...

**Comment by klonos** [comment]
Hmm, I am leaning towards what @BWPanda is suggesting at the moment. If I am installing a module/theme/layout from the admin UI of a specific site, then I expect that to be installed/available only for that site.

Having said that, I see the reason why some might want it to be done the other way (installed globally, for all sites), but cannot think of any way to make this more intuitive ...unless we provide an option in the Installer UI.

**Comment by docwilmot** [comment]
I'm going to change this from being a bug, and leave it open for now. Maybe call it a feature request.

**Comment by klonos** [comment]
Yeah, I think that how this would work ideally is if there was something like an "Install for all sites" checkbox somewhere in the installer steps, and then have the installer extract files to the proper dir. Depends on the code/logic complexity if this should be in core, or contrib.

**Comment by BWPanda** [comment]
I don't see how this is a contributed

**Comment by klonos** [comment]
@BWPanda as I said, this sounds like a very interesting problem to solve, and it would save work/time for people, so UX++ (site builders). I may take a crack at it ...if either you or anyone else don't beat me to it 🙂

**Comment by BWPanda** [comment]
Here's a PR: https://github.com/backdrop/backdrop/pull/3946

And a Lando file for testing can be found here: https://github.com/backdrop/backdrop-issues/issues/4151

**Comment by klonos** [comment]
Yup, we should simplify, so that it makes sense to less technical people. Perhaps something that says:

> Multisite setup detected. Make these projects available for:
> ( ) all sites in this multisite setup
> ( ) only for the current site

Perhaps also denote the actual file paths? (as help text to each individual option)

> Update: wait, isn't this also for themes and layouts?

Right, we'd need to account for all that.

**Comment by klonos** [comment]
...I just recalled some very good advise from @olafgrabienski on another thread re making sure that the labels of fields and the respective options are all individual sentences that can stand on their own. So my previous suggestion would need to be tweaked accordingly. I don't have a good suggestion at this time, but you get the point 😉

**Comment by klonos** [comment]
...this is as good as I could get it to:

> Multisite setup detected
> ( ) Make these projects available for all sites in this multisite setup
> ( ) Make these projects available only for the current site

**Comment by BWPanda** [comment]
I originally considered two ways to word this:
1. *Where* the project would be downloaded (e.g. root directory, site-specific directory)
2. *What* sites the project would be available for (e.g. all sites, this site)

I went with the first one, but I'm happy to try the second.

> Yup, we should simplify, so that it makes sense to less technical people.

I'm happy to try different wording, but I disagree with that premise. This field is only shown for multisites, and if someone is using a multisite then they would know the difference between the site-specific directory and the root directory. You could argue that someone might just be managing a site that was setup by someone else (e.g. a site admin with no access to the file system), but I still think they'd need to know the difference, especially when they can be installing modules that will be available to other sites (which they might not manage).

> I just recalled some very good advise [...] re making sure that the labels of fields and the respective options are all individual sentences that can stand on their own.

I also disagree with this. Examples from core that don't do this:

![image](https://user-images.githubusercontent.com/2385329/153345054-2626a017-7e36-4280-8e6d-9b0547a5a738.png)

![image](https://user-images.githubusercontent.com/2385329/153345131-2355ba8a-ca84-4446-b3ed-2fb2e5223c6d.png)

![image](https://user-images.githubusercontent.com/2385329/153345156-cb206bce-8c24-4be0-932d-c5c9189fc13b.png)

If the plan is to ultimately change those too, I think that's just going to unnecessarily duplicate words in the UI. I mean, that's what field labels are for: giving context to the options so we don't have to repeat the same thing for each one. If we don't plan to change the above examples, then I don't see the point of changing this PR either.

I've updated the PR to add a description to each option with the path in question. If we want to try the 2nd wording as described above, then I don't think a path in the description makes sense then...

Now looks like this:

![image](https://user-images.githubusercontent.com/2385329/153347436-1f7ea7c4-a39b-4dcd-b197-7dc6e9f75e8f.png)

**Comment by indigoxela** [comment]
TBH, I actually liked the direction of @klonos' suggestion, though not ideal (yet). But I usually can't help much with wording.

And I still find the word "Root" misleading and rather confusing (even for the "slightly more technical user"). :wink:

**Comment by olafgrabienski** [comment]
>> making sure that the labels of fields and the respective options are all individual sentences that can stand on their own

> I also disagree with this. Examples from core that don't do this (...)

@BWPanda I agree the examples shouldn't be changed but they don't match what I [had in mind](https://github.com/backdrop/backdrop-issues/issues/4253#issuecomment-1008326477). I don't think, labels and options have to be individual sentences. However, *if* there are sentences,

> a sentence shouldn't be split up in two parts (label and option), so that the option isn't descriptive without the label

Incomplete sentences like `Download project(s) to the appropriate folder in the` are hard to understand for people translating Backdrop strings into other languages.

Example for an alternative:

`Parent directory for download project(s):`
`- Root / Shared`
`- Site-specific`

**Comment by BWPanda** [comment]
Thanks @olafgrabienski, that makes more sense.

**Comment by klonos** [comment]
Yes, what @olafgrabienski said ^^ ...sorry I failed to explain things properly. That's what I meant.

**Comment by BWPanda** [comment]
PR updated. How's this?

![image](https://user-images.githubusercontent.com/2385329/153388761-0ba72215-a4a0-4def-99ed-9cc762407a24.png)

**Comment by indigoxela** [comment]
> How's this?

A lot better!

One thought: Is the "/app/www" part of the example path actually helpful? How about "/modules/" and "/sites/example.com/modules/"?

**Comment by klonos** [comment]
Yup, very much better @BWPanda 👍🏼  ...and I like @indigoxela's suggestion too 👍🏼  (changing the `/multisite/` bit to `/example.com/` - don't have any strong feelings re removing the `/app/www/` bits).

**Comment by BWPanda** [comment]
> Is the "/app/www" part of the example path actually helpful?

So that's dynamically generated based on the actual site. The root path is obtained using `BACKDROP_ROOT`, while the site-specific path uses `conf_path()`. So this'll show the user the actual path to their actual directories.

The reason for the `/app/www/` part is that it was included by default (for me) with `BACKDROP_ROOT`. Whereas `conf_path()` was just `./sites/multisite`. It looked weird having one path be absolute and one be relative, so that's why I wrapped them both in `realpath()`.

**Comment by klonos** [comment]
> So that's dynamically generated based on the actual site. The root path is obtained using `BACKDROP_ROOT`, while the site-specific path uses `conf_path()`. So this'll show the user the actual path to their actual directories.

In that case (and sorry I didn't check the actual code before commenting), I believe that it is indeed worth having those there 👍🏼

**Comment by klonos** [comment]
>...but should we then remove the `e.g.`s, since these are actual paths and not examples?

...or replace them with `i.e.`s instead, which is more accurate?

**Comment by BWPanda** [comment]
The reason I used `e.g.` is because the `/modules` part is hard-coded.

**Comment by klonos** [comment]
Right 👍🏼 ...I missed that.

**Comment by BWPanda** [comment]
So lots of feedback here, but anyone actually want to test/review? Or if you have already, tag it as such?

**Comment by yorkshire-pudding** [comment]
@BWPanda - I've tested this.  If the modules folder has not yet been created in the sites/multi_one/ folder then the UI cannot install the module.

This is the error I get:
```
Error installing Paragraphs. Error installing / updating. Error: Unable to create <em class="placeholder"></em> due to the following: <em class="placeholder">Cannot chmod &lt;em class=&quot;placeholder&quot;&gt;&lt;/em&gt;.</em>
```
PS - this is exactly what is rendered on screen - this is not an error with copying.  It is exactly the same in the log.

Once `bee` has installed a module in that sites directory (or I manually create the modules folder in that location), then the UI can install there.

**Comment by docwilmot** [comment]
I was trying to resurrect this old issue
> @BWPanda - I've tested this. If the modules folder has not yet been created in the sites/multi_one/ folder then the UI cannot install the module.

I can confirm this error. We can get rid of the error by automatically creating the project (module/theme/layout) folder during download. But that raises the question: is there a technical/security reason why multisites must be manual? If there is, then we shouldnt fix this by creating the folder. 

But also, if the user has to have access to the server to create the multisite folder, then its one further step to just also create a project folder manually, so just requiring it to be manual doesnt add much work.

**Comment by docwilmot** [comment]
New PR to replace original since @BWPanda repo doesnt exist anymore.

**Comment by robertgarrigos** [comment]
I tested this PR, and it works as expected. It creates the needed directories, also. I would love to see this PR applied, as I would use it A LOT. My only concern is that I would like to be able, as admin of the main site, to prohibit the subsites from installing in the main modules directory. But maybe this could be done with a submodule?
